src/syscall.rs

```diff
@@ -331,19 +341,96 @@ impl<'c> SyscallContext<'c> {
 
         let key = u256_from_u128(stg_key.hi, stg_key.lo);
 
-        let result = self.db.read_storage(address, key);
+        // Read value from journaled_storage. If there isn't one, then read from db
+        let result = self
+            .inner_context
+            .journaled_storage
+            .get(&key)
+            .map(|slot| slot.present_value)
+            .unwrap_or(self.db.read_storage(address, key));
```

Comment:
We use `*_or_else` functions when computation should be done lazily (like here).
```suggestion
            .unwrap_or_else(|| self.db.read_storage(address, key));
```